### PR TITLE
Sync OTEL config volume/mount and args

### LINKED ISSUE #1266: OTel agent/collector failing with duplicate volume/volumeMount
Using the operator to deploy the OTel variants of the Jaeger agent/collector requires adjusting the `health_check` port, which is done by supplying `config`. Minimal example from a minikube deployment (reproduced after initially hitting the issue in a full kubernetes cluster):

```
apiVersion: jaegertracing.io/v1
kind: Jaeger
metadata:
  name: simplest
spec:
  strategy: production
  agent:
    image: jaegertracing/jaeger-opentelemetry-agent:latest
    strategy: DaemonSet
    config:
      extensions:
        health_check:
          port: 14271
  collector:
    image: jaegertracing/jaeger-opentelemetry-collector:latest
    config:
      extensions:
        health_check:
          port: 14269
  storage:
    type: elasticsearch
    options:
      es:
        server-urls: http://jaeger-es-http:9200
```

With this configuration in place, the operator produces a _duplicate_ `volume` and `volumeMount` for the otel config, and thus fails reconciliation repeatedly:

```
time="2020-10-20T01:46:51Z" level=error msg="failed to apply the changes" error="Deployment.apps \"simplest-query\" is invalid: spec.template.spec.containers[1].volumeMounts[1].mountPath: Invalid value: \"/etc/jaeger/otel/\": must be unique" execution="2020-10-20 01:46:51.030423915 +0000 UTC" instance=simplest namespace=observability
time="2020-10-20T01:46:52Z" level=error msg="failed to apply the changes" error="Deployment.apps \"simplest-query\" is invalid: spec.template.spec.containers[1].volumeMounts[1].mountPath: Invalid value: \"/etc/jaeger/otel/\": must be unique" execution="2020-10-20 01:46:52.101401604 +0000 UTC" instance=simplest namespace=observability
time="2020-10-20T01:46:53Z" level=error msg="failed to apply the changes" error="Deployment.apps \"simplest-query\" is invalid: spec.template.spec.containers[1].volumeMounts[1].mountPath: Invalid value: \"/etc/jaeger/otel/\": must be unique" execution="2020-10-20 01:46:53.187396068 +0000 UTC" instance=simplest namespace=observability
time="2020-10-20T01:46:54Z" level=error msg="failed to apply the changes" error="Deployment.apps \"simplest-query\" is invalid: spec.template.spec.containers[1].volumeMounts[1].mountPath: Invalid value: \"/etc/jaeger/otel/\": must be unique" execution="2020-10-20 01:46:54.272854654 +0000 UTC" instance=simplest namespace=observability
time="2020-10-20T01:46:55Z" level=error msg="failed to apply the changes" error="Deployment.apps \"simplest-query\" is invalid: spec.template.spec.containers[1].volumeMounts[1].mountPath: Invalid value: \"/etc/jaeger/otel/\": must be unique" execution="2020-10-20 01:46:55.359538151 +0000 UTC" instance=simplest namespace=observability
```

This appears to be as a result of a few factors:

* The operator state has the agent config saved back into it from previous runs
* The operator state is merged together including the config `volume`/`volumeMount` to prepare the agent
* `otelconfig.ShouldCreate` returns `true` when 'agent options' are empty, but 'agent config' is not empty, triggering `otelconfig.Update`, which adds the duplicate `volume`/`volumeMount` (first time through I assume this was fine, because the operator state didn't also have it?)
* Any de-duplication of volumes/volumeMounts occurs _before_ the OTel config is handled (during merge/CA handling)

Operator state:

`kubectl -n observability get jaegers.jaegertracing.io -oyaml`

```
...
  spec:
    agent:
      image: jaegertracing/jaeger-opentelemetry-agent:latest
      options: {}
      resources: {}
      strategy: DaemonSet
      volumeMounts:
      - mountPath: /etc/jaeger/otel/
        name: simplest-agent-otel-config
        readOnly: true
      volumes:
      - configMap:
          defaultMode: 420
          items:
          - key: config
            path: config.yaml
          name: simplest-agent-otel-config
        name: simplest-agent-otel-config
...
```

### PR DESCRIPTION
**Pull request by jpkrohling**
Fixes #1266

﻿Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

### CODE REVIEW

**Comment by dackroyd** [comment]
This has addressed the duplicate `volume`/`volumeMount` with the agent, which allows the agent to be updated correctly.

With the current implementation in this PR, the `volume` and `volumeMount` can never be _removed_ once they have been created, i.e if the `config` needs to be removed at a later stage, the volume/volumeMount will not be removed (since they are merged in from the operator spec)

File: pkg/config/otelconfig/otelconfig.go [hunk 1]
```diff
@@ -106,7 +104,10 @@ func create(jaeger *v1.Jaeger, component string, otelConfig map[string]interface
 
 // Update injects required flags and objects to the common spec.
 func Update(jaeger *v1.Jaeger, component string, commonSpec *v1.JaegerCommonSpec, args *[]string) {
-	volume := corev1.Volume{
+	configFileLocation := fmt.Sprintf("/etc/jaeger/otel/%s/", component)
```
Comment (pavolloffay) [inline pkg/config/otelconfig/otelconfig.go:107]: nit: is there a to fix this without embedding the component name in the path? I find it redundant but I might be missing something.

Comment (jpkrohling) [reply pkg/config/otelconfig/otelconfig.go]: I reverted this change, at first I thought it was part of the problem but it actually wasn't.

**Comment by jpkrohling** [comment]
@dackroyd, this should now address the case where the OTEL configuration is removed (or an external file is now specified). An updated image was published at the same location, in case you want to give this a try: `quay.io/jpkroehling/jaeger-operator:pr1268`.

File: pkg/config/otelconfig/otelconfig_test.go [hunk 2]
```diff
@@ -81,10 +80,74 @@ func TestUpdate(t *testing.T) {
 	j := v1.NewJaeger(types.NamespacedName{Name: "jaeger"})
 	args := []string{}
 	commonSpec := &v1.JaegerCommonSpec{}
-	Update(j, "agent", commonSpec, &args)
-	assert.Equal(t, []string{configFlagWithFile}, args)
+	upsert(j, "agent", commonSpec, &args)
+	assert.Equal(t, []string{"--config=/etc/jaeger/otel/config.yaml"}, args)
```
Comment (objectiser) [inline pkg/config/otelconfig/otelconfig_test.go:84]: Return to using `configFlagWithFile`.

File: pkg/config/otelconfig/otelconfig_test.go [hunk 3]
```diff
@@ -81,10 +80,74 @@ func TestUpdate(t *testing.T) {
...
 	assert.Equal(t, "jaeger-agent-otel-config", commonSpec.Volumes[0].Name)
 	assert.Equal(t, "jaeger-agent-otel-config", commonSpec.VolumeMounts[0].Name)
 	assert.Equal(t, []corev1.KeyToPath{{Key: "config", Path: configFileName}}, commonSpec.Volumes[0].ConfigMap.Items)
-	assert.Equal(t, configFileLocation, commonSpec.VolumeMounts[0].MountPath)
+	assert.Equal(t, "/etc/jaeger/otel/", commonSpec.VolumeMounts[0].MountPath)
```
Comment (objectiser) [inline pkg/config/otelconfig/otelconfig_test.go:88]: Revert to using `configFileLocation`.

File: pkg/config/otelconfig/otelconfig_test.go [hunk 4]
```diff
@@ -81,10 +80,74 @@ func TestUpdate(t *testing.T) {
...
+	assert.Len(t, commonSpec.Volumes, 2)
+	assert.Nil(t, commonSpec.Volumes[1].VolumeSource.HostPath)
+	assert.NotEqual(t, "/existing/stale/configuration", commonSpec.VolumeMounts[1].MountPath)
+	assert.Len(t, args, 2)
+}
+
+func TestSyncShouldRemove(t *testing.T) {
+	// prepare
+	j := v1.NewJaeger(types.NamespacedName{Name: "jaeger"})
+	args := []string{"--some-unrelated-arg"}
```
Comment (objectiser) [inline pkg/config/otelconfig/otelconfig_test.go:132]: Shouldn't the `configFlagWithFile` arg also be included here, to make sure removed?

**Review by objectiser — APPROVED** [review]

**Pull request merged**